### ISSUE TITLE
Segfault when using QueryRowContext

### ISSUE THREAD

**Issue by VadimIpatov** [godror/godror#100]
**Describe the bug**
Godror panics if any connection problem happens during doing a query with context.WithTimeout.

```
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x80 pc=0x31aa1338]

runtime stack:
runtime.throw(0x41b626f, 0x2a)
	/usr/local/go/src/runtime/panic.go:1114 +0x72
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:679 +0x46a

goroutine 21 [syscall]:
runtime.cgocall(0x415ded0, 0xc000145818, 0xc000145828)
	/usr/local/go/src/runtime/cgocall.go:133 +0x5b fp=0xc0001457e8 sp=0xc0001457b0 pc=0x400548b
github.com/godror/godror._Cfunc_dpiStmt_release(0xce00050, 0x0)
	_cgo_gotypes.go:4462 +0x49 fp=0xc000145818 sp=0xc0001457e8 pc=0x40e6d79
github.com/godror/godror.(*statement).closeNotLocking.func2(0xce00050, 0x0)
	/Users/vadimipatov/dev/golang/pkg/mod/github.com/godror/godror@v0.19.4/stmt.go:296 +0x4d fp=0xc000145848 sp=0xc000145818 pc=0x4114ead
github.com/godror/godror.(*statement).closeNotLocking(0xc0001b03c0, 0x0, 0x0)
	/Users/vadimipatov/dev/golang/pkg/mod/github.com/godror/godror@v0.19.4/stmt.go:296 +0x249 fp=0xc0001458e0 sp=0xc000145848 pc=0x40fc2d9
github.com/godror/godror.(*statement).queryContextNotLocked(0xc0001b03c0, 0x41e3020, 0xc00019e600, 0x42fa2c8, 0x0, 0x0, 0x0, 0x8, 0x0, 0x0)
	/Users/vadimipatov/dev/golang/pkg/mod/github.com/godror/godror@v0.19.4/stmt.go:606 +0x6ac fp=0xc000145ac0 sp=0xc0001458e0 pc=0x40febcc
github.com/godror/godror.(*statement).QueryContext(0xc0001b03c0, 0x41e3020, 0xc00019e600, 0x42fa2c8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/vadimipatov/dev/golang/pkg/mod/github.com/godror/godror@v0.19.4/stmt.go:510 +0x152 fp=0xc000145b50 sp=0xc000145ac0 pc=0x40fe3c2
database/sql.ctxDriverStmtQuery(0x41e3020, 0xc00019e600, 0x41e30a0, 0xc0001b03c0, 0x42fa2c8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/usr/local/go/src/database/sql/ctxutil.go:82 +0x210 fp=0xc000145bc8 sp=0xc000145b50 pc=0x40b9c50
database/sql.rowsiFromStatement(0x41e3020, 0xc00019e600, 0x41e2960, 0xc00013c000, 0xc0001bc3c0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/usr/local/go/src/database/sql/sql.go:2604 +0x16a fp=0xc000145c48 sp=0xc000145bc8 pc=0x40bee9a
database/sql.(*DB).queryDC(0xc0001be000, 0x41e3020, 0xc00019e600, 0x0, 0x0, 0xc000076000, 0xc0001813b0, 0x41b0bc6, 0x12, 0x0, ...)
	/usr/local/go/src/database/sql/sql.go:1622 +0x230 fp=0xc000145e08 sp=0xc000145c48 pc=0x40be700
database/sql.(*DB).query(0xc0001be000, 0x41e3020, 0xc00019e600, 0x41b0bc6, 0x12, 0x0, 0x0, 0x0, 0x41e2f01, 0xc00018a000, ...)
	/usr/local/go/src/database/sql/sql.go:1570 +0x136 fp=0xc000145e98 sp=0xc000145e08 pc=0x40be436
database/sql.(*DB).QueryContext(0xc0001be000, 0x41e3020, 0xc00019e600, 0x41b0bc6, 0x12, 0x0, 0x0, 0x0, 0xc0001b4180, 0x0, ...)
	/usr/local/go/src/database/sql/sql.go:1547 +0xd1 fp=0xc000145f28 sp=0xc000145e98 pc=0x40be171
database/sql.(*DB).QueryRowContext(...)
	/usr/local/go/src/database/sql/sql.go:1648
main.main.func1(0xc0001a4050, 0xc0001be000)
	/Users/vadimipatov/dev/tmp/godor-test/app.go:41 +0xf8 fp=0xc000145fd0 sp=0xc000145f28 pc=0x411d698
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc000145fd8 sp=0xc000145fd0 pc=0x40616a1
created by main.main
	/Users/vadimipatov/dev/tmp/godor-test/app.go:28 +0x116

goroutine 1 [select (no cases)]:
main.main()
	/Users/vadimipatov/dev/tmp/godor-test/app.go:60 +0x11b

goroutine 19 [select]:
database/sql.(*DB).connectionOpener(0xc0001be000, 0x41e2fa0, 0xc0001bc000)
	/usr/local/go/src/database/sql/sql.go:1052 +0xe8
created by database/sql.OpenDB
	/usr/local/go/src/database/sql/sql.go:722 +0x15d

goroutine 20 [select]:
database/sql.(*DB).connectionResetter(0xc0001be000, 0x41e2fa0, 0xc0001bc000)
	/usr/local/go/src/database/sql/sql.go:1065 +0xfb
created by database/sql.OpenDB
	/usr/local/go/src/database/sql/sql.go:723 +0x193
exit status 2
```

**To Reproduce**
1. run this code:
```go
package main

import (
	"context"
	"database/sql"
	"log"
	"time"

	_ "github.com/godror/godror"	
)

func main() {
	db, err := sql.Open("godror", `user="SYSTEM" password="password" connectString="localhost:1521/XE"`)
	if err != nil {
		log.Printf("Error: %s", err.Error())
	}

	ticker := time.NewTicker(1 * time.Second)

	go func() {
		for {
			select {
				case <- ticker.C:
					ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
					defer cancel()

					row := db.QueryRowContext(ctx, "SELECT 1 FROM DUAL")
					if ctx.Err() == context.DeadlineExceeded {
						log.Printf("Error: Timeout")
						continue
					}					

					var res int
					err := row.Scan(&res)

					if err != nil {
						log.Printf("Error: %s", err.Error())
						continue
					}

					log.Printf("Result: %d", res)
			}
		}
	}()
	
	select {}
}

```

2. disconnect a network cable or stop database
3. wait for failure in 10-15 seconds

**Expected behavior**
Keep calm and don't panic :) Act like "github.com/lib/pq"

**Your oracle client version**
19.3.0.0.0

**Your godror version**
v0.19.4

**Your go version**
go version go1.14.1 darwin/amd64

**Your gcc version**
Apple LLVM version 10.0.1 (clang-1001.0.46.3)

**Machine (please complete the following information):**
 - OS: MacOs
 - Architecture x86_64
 - Version: 10.14

**Additional context**
PingContext works fine.

**Comment by tgulacsi** [comment]
PTAL e36a83e - it may manage only stmt.closeNotLocking, not all possible panics!

pq has an easier route: no cgo.

**Comment by VadimIpatov** [comment]
> PTAL [e36a83e](https://github.com/godror/godror/commit/e36a83e1630fc7a5ea9565841a997426b10ac9ba) - it may manage only stmt.closeNotLocking, not all possible panics!

I have more stack traces, probably related to this problem (maybe not), from my real project:

- [panic1](https://github.com/godror/godror/files/5236618/oracle_panic1.txt)
- [panic2](https://github.com/godror/godror/files/5236620/oracle_panic2.txt)
- [panic3](https://github.com/godror/godror/files/5236621/oracle_panic3.txt)
- [panic4](https://github.com/godror/godror/files/5236622/oracle_panic4.txt)

I hope it can help.

**Comment by tgulacsi** [comment]
panic3: I'll add recover to acquireConn.
panic1, panic2, panic4: It's already recovered, so I don't know what more can we do.

**Comment by VadimIpatov** [comment]
> panic1, panic2, panic4: It's already recovered, so I don't know what more can we do.
do you mean it's fixed in e36a83e?

**Comment by tgulacsi** [comment]
Re https://github.com/godror/godror/issues/100#issuecomment-694179629 : no, it's been fixed since about v0.16.1
I don't know what else could I do - I already recover from the panic.
As https://stackoverflow.com/questions/43212593/handling-sigsegv-with-recover suggests, SIGSEGV crashes the program.
I'll try debug.SetPanicOnFault.

PTAL 2dab250

**Comment by VadimIpatov** [comment]
> PTAL [2dab250](https://github.com/godror/godror/commit/2dab250ab19e158ba97699841f40c9de9d061f29)

I've tried. It still panics:

```
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x80 pc=0x307fd338]

runtime stack:
runtime.throw(0x41b785a, 0x2a)
	/usr/local/go/src/runtime/panic.go:1114 +0x72
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:679 +0x46a

goroutine 84 [syscall]:
runtime.cgocall(0x415f370, 0xc00003d758, 0x0)
	/usr/local/go/src/runtime/cgocall.go:133 +0x5b fp=0xc00003d728 sp=0xc00003d6f0 pc=0x4004f1b
github.com/godror/godror._Cfunc_dpiStmt_release(0xd202b40, 0x0)
	_cgo_gotypes.go:4479 +0x49 fp=0xc00003d758 sp=0xc00003d728 pc=0x40e76f9
github.com/godror/godror.(*statement).closeNotLocking.func2.2(0xd202b40, 0xc000090700)
	/Users/vadimipatov/dev/golang/pkg/mod/github.com/godror/godror@v0.20.1-0.20200917120015-2dab250ab19e/stmt.go:306 +0x4d fp=0xc00003d788 sp=0xc00003d758 pc=0x411619d
github.com/godror/godror.(*statement).closeNotLocking.func2(0xd202b40, 0xc0000908a0)
	/Users/vadimipatov/dev/golang/pkg/mod/github.com/godror/godror@v0.20.1-0.20200917120015-2dab250ab19e/stmt.go:306 +0x74 fp=0xc00003d7d0 sp=0xc00003d788 pc=0x4116234
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc00003d7d8 sp=0xc00003d7d0 pc=0x4061291
created by github.com/godror/godror.(*statement).closeNotLocking
	/Users/vadimipatov/dev/golang/pkg/mod/github.com/godror/godror@v0.20.1-0.20200917120015-2dab250ab19e/stmt.go:297 +0x28a

goroutine 1 [select (no cases)]:
main.main()
	/Users/vadimipatov/dev/tmp/godor-test/app.go:61 +0x11b

goroutine 6 [select]:
database/sql.(*DB).connectionOpener(0xc000122000, 0x41e4c40, 0xc000026080)
	/usr/local/go/src/database/sql/sql.go:1052 +0xe8
created by database/sql.OpenDB
	/usr/local/go/src/database/sql/sql.go:722 +0x15d

goroutine 7 [select]:
database/sql.(*DB).connectionResetter(0xc000122000, 0x41e4c40, 0xc000026080)
	/usr/local/go/src/database/sql/sql.go:1065 +0xfb
created by database/sql.OpenDB
	/usr/local/go/src/database/sql/sql.go:723 +0x193

goroutine 8 [chan receive]:
github.com/godror/godror.(*statement).closeNotLocking(0xc0000bc280, 0x0, 0x0)
	/Users/vadimipatov/dev/golang/pkg/mod/github.com/godror/godror@v0.20.1-0.20200917120015-2dab250ab19e/stmt.go:309 +0x2b0
github.com/godror/godror.(*statement).queryContextNotLocked(0xc0000bc280, 0x41e4cc0, 0xc0000907e0, 0x42fd328, 0x0, 0x0, 0x0, 0x8, 0x0, 0x0)
	/Users/vadimipatov/dev/golang/pkg/mod/github.com/godror/godror@v0.20.1-0.20200917120015-2dab250ab19e/stmt.go:625 +0x6ac
github.com/godror/godror.(*statement).QueryContext(0xc0000bc280, 0x41e4cc0, 0xc0000907e0, 0x42fd328, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/vadimipatov/dev/golang/pkg/mod/github.com/godror/godror@v0.20.1-0.20200917120015-2dab250ab19e/stmt.go:527 +0x152
database/sql.ctxDriverStmtQuery(0x41e4cc0, 0xc0000907e0, 0x41e4d40, 0xc0000bc280, 0x42fd328, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/usr/local/go/src/database/sql/ctxutil.go:82 +0x210
database/sql.rowsiFromStatement(0x41e4cc0, 0xc0000907e0, 0x41e4580, 0xc0001aa000, 0xc0000a8440, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/usr/local/go/src/database/sql/sql.go:2604 +0x16a
database/sql.(*DB).queryDC(0xc000122000, 0x41e4cc0, 0xc0000907e0, 0x0, 0x0, 0xc0001ce000, 0xc000095100, 0x41b20f8, 0x12, 0x0, ...)
	/usr/local/go/src/database/sql/sql.go:1622 +0x230
database/sql.(*DB).query(0xc000122000, 0x41e4cc0, 0xc0000907e0, 0x41b20f8, 0x12, 0x0, 0x0, 0x0, 0x41e4c01, 0xc000018070, ...)
	/usr/local/go/src/database/sql/sql.go:1570 +0x136
database/sql.(*DB).QueryContext(0xc000122000, 0x41e4cc0, 0xc0000907e0, 0x41b20f8, 0x12, 0x0, 0x0, 0x0, 0xc00011c180, 0x0, ...)
	/usr/local/go/src/database/sql/sql.go:1547 +0xd1
database/sql.(*DB).QueryRowContext(...)
	/usr/local/go/src/database/sql/sql.go:1648
main.main.func1(0xc0000720a0, 0xc000122000)
	/Users/vadimipatov/dev/tmp/godor-test/app.go:42 +0xf8
created by main.main
	/Users/vadimipatov/dev/tmp/godor-test/app.go:29 +0x116
exit status 2
```

**Comment by tgulacsi** [comment]
I don't have any more idea.

**Comment by tgulacsi** [comment]
I've tried asking for help: https://groups.google.com/g/golang-nuts/c/j4npkSTUZFs

**Comment by sudarshan12s** [comment]
It looks like dpiStmtRelease is done twice.
one triggered from db.QueryRowContext causes  closeNotLocking resulted from
line stm.go:461 which does dpiStmtRelease

other one from line row.Scan(&res) , which does (*rows).Close and
 dpiStmtRelease.

 case <-ctx.Done():
 456             err := ctx.Err()
 457             if Log != nil {
 458                 Log("msg", "BREAK statement", "error", err)
 459             }
 460             _ = st.Break()
 461             st.closeNotLocking()
 462             return nil, err
 463         }
 464     }


somehow reproducing for me is little tough, i need to change IP of the mac
os hosting the oracle DB container. its not coming always.  hence i
cant easily change the  application program to try other possibilities.

**Comment by tgulacsi** [comment]
@sudarshan12s thanks for the help!

But st.Break() is just st.conn.Break(), which calls dpiConn_breakExecution(dpiConn).
Does THAT call dpiStmtRelease somewhere? Because st.Break does not release the dpiStmt explicitly.

In 301565b I deleted the ineffective recover codes, simplified it - the Go side cannot catch the SIGSEGV of the C side.

**Comment by sudarshan12s** [comment]
@tgulacsi,
yes st.Break doesn't cause dpiStmtRelease .

I cross checked if dpiStmtRelease or dpiStmtExecute improper order can
cause segfaults , but it was returning with proper error.  so mostly issue
is something else here.

still figuring out easy way to reproduce the issue for RCA

**Comment by sudarshan12s** [comment]
@vadimIpatov <https://github.com/VadimIpatov>,
Can you increase the context deadline and see if the issue happens still?

I think issue could be the way the C.dpiStmt_execute is run inside a
goroutine for handling context deadline.

stmt.go:578, main goroutine  waits  on select for another goroutine to
complete C.dpiStmt_execute . But before it could finish the main goroutine
calls
_ = st.Break() as deadline exceeded.

The goroutine spawned could be stuck at dpiStmt_execute and main goroutine
would proceed to release this stmt and connection.

checking more on these ...

**Comment by tgulacsi** [comment]
@anthony-tuininga helped to reveal the error: after calling dpiConn_breakExecution, NOTHING should be called till the executing statement returns with error - esp. not the close(NotLocking).


If the context deadline is reached, a call to dpiConn_breakExecution() should be made in order to actually terminate the call to dpiStmt_execute() -- which should then fail with ORA-1013: user requested cancel of current operation. This needs to be done in a separate thread. You then need to wait for the call to dpiStmt_execute() to complete with the error before proceeding.

Here is an example that may be easier to follow:

```
create or replace type T_ISSUE_100_ROWS as table of TIMESTAMP;
/

create or replace FUNCTION F_ISSUE_100 (
    ITERS IN VARCHAR2 DEFAULT 10,
    PAUSE IN VARCHAR2 DEFAULT 1
) RETURN T_ISSUE_100_ROWS AUTHID CURRENT_USER PIPELINED AS
BEGIN
    FOR i IN 1..ITERS LOOP
        DBMS_SESSION.SLEEP(PAUSE);
        PIPE ROW(SYSTIMESTAMP);
    END LOOP;
END;
/
```

This creates a view that will only fetch rows after a period of time. You can set the prefetchRows attribute to specify how many it will fetch during the call to dpiStmt_execute(). The default value is 2.

Then you can create a go program like the following:

```
package main

import (
    "context"
    "database/sql"
    "log"
    "time"

    _ "github.com/godror/godror"
)

func main() {
    const dsn = `user="cx_oracle_bugs" password="welcome" connectString="localhost/DB21PDB1"`
    const statement = "SELECT * FROM TABLE(F_ISSUE_100(iters => 15, pause => 1))"

    db, err := sql.Open("godror", dsn)
    if err != nil {
        log.Printf("Error: %s", err.Error())
    }

    ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
    defer cancel()

    row := db.QueryRowContext(ctx, statement)

    if ctx.Err() == context.DeadlineExceeded {
        log.Printf("Error: Timeout")
    }

    var res string
    err = row.Scan(&res)
    if err != nil {
        log.Printf("Row Fetch Error: %s", err.Error())
    }

    log.Printf("Result: %s", res)
}
```

If you set `DPI_DEBUG_LEVEL` to level 6 you will see the following:

```
ODPI [46190] 2020-09-18 15:15:01.291: fn start dpiStmt_execute(0x1421a80)
ODPI [46192] 2020-09-18 15:15:03.233: fn start dpiConn_breakExecution(0x140ff80)
ODPI [46192] 2020-09-18 15:15:03.233: fn end dpiConn_breakExecution(0x140ff80) -> 0
ODPI [46192] 2020-09-18 15:15:03.233: fn start dpiStmt_release(0x1421a80)
ODPI [46192] 2020-09-18 15:15:03.233: ref 0x1421a80 (dpiStmt) -> 0
ODPI [46192] 2020-09-18 15:15:03.233: ref 0x140ff80 (dpiConn) -> 1
ODPI [46192] 2020-09-18 15:15:03.233: fn end dpiStmt_release(0x1421a80) -> 0
ODPI [46190] 2020-09-18 15:15:03.233: fn end dpiStmt_execute(0x1421a80) -> -1
ODPI [46192] 2020-09-18 15:15:03.233: fn start dpiConn_release(0x140ff80)
ODPI [46192] 2020-09-18 15:15:03.233: ref 0x140ff80 (dpiConn) -> 0
ODPI [46192] 2020-09-18 15:15:03.234: ref 0x1389390 (dpiPool) -> 1
ODPI [46192] 2020-09-18 15:15:03.234: fn end dpiConn_release(0x140ff80) -> 0
```

Note that the call to dpiStmt_execute is running in thread 46190. Two seconds after it starts, thread 46192 calls dpiConn_breakExecution(). It also (apparently) calls dpiStmt_release() and dpiConn_release() at this point! In between you see that thread 46190 receives the error (return value of -1). Whatever the cancel is doing it needs to *wait* after calling dpiConn_breakExecution() until the dpiStmt_execute() returns with an error before continuing on.

Hope that explains things!

Anthony

**Comment by tgulacsi** [comment]
It should be fixed with v0.20.1.

**Comment by VadimIpatov** [comment]
> It should be fixed with v0.20.1.

confirmed. the problem is fixed.
thanks a lot!
